smart-contracts/contracts/cl-vault/src/vault/deposit.rs

```diff
@@ -36,90 +41,218 @@ pub(crate) fn _execute_any_deposit(
     unimplemented!()
 }
 
-/// Try to deposit as much user funds as we can into the a position and
-/// refund the rest to the caller
+/// allocate as many user funds as we can to the current positions, after adding to those positions
 pub(crate) fn execute_exact_deposit(
-    deps: DepsMut,
+    mut deps: DepsMut,
     env: Env,
     info: MessageInfo,
     recipient: Option<String>,
 ) -> Result<Response, ContractError> {
     // Unwrap recipient or use caller's address
     let recipient = recipient.map_or(Ok(info.sender.clone()), |x| deps.api.addr_validate(&x))?;
+    CURRENT_DEPOSITOR.save(deps.storage, &recipient)?;
 
-    let position_id = (POSITION.load(deps.storage)?).position_id;
-    let position = ConcentratedliquidityQuerier::new(&deps.querier)
-        .position_by_id(position_id)?
-        .position
-        .ok_or(ContractError::PositionNotFound)?
-        .position
-        .ok_or(ContractError::PositionNotFound)?;
+    debug!(deps, "depositing 1");
+    let positions = get_positions(deps.storage, &deps.querier)?;
+
+    debug!(deps, "depositing 2");
 
     let pool = POOL_CONFIG.load(deps.storage)?;
     let (token0, token1) = must_pay_one_or_two(&info, (pool.token0, pool.token1))?;
 
-    CURRENT_DEPOSIT.save(
-        deps.storage,
-        &CurrentDeposit {
-            token0_in: token0.amount,
-            token1_in: token1.amount,
-            sender: recipient,
-        },
+    let spot_price = get_spot_price(deps.storage, &deps.querier)?;
+
+    let psf = allocate_funds_per_position(
+        deps.branch(),
+        positions.clone(),
+        spot_price,
+        token0.amount,
+        token1.amount,
     )?;
 
-    // Create coins_to_send with no zero amounts
-    let mut coins_to_send = vec![];
-    if !token0.amount.is_zero() {
-        coins_to_send.push(token0.clone());
-    }
-    if !token1.amount.is_zero() {
-        coins_to_send.push(token1.clone());
-    }
+    let position_funds = psf.iter().zip(positions);
```

Comment:
Could this be optmized in some way, as type of position_funds would become,  Zip<Iter<(Position, Uint128, Uint128)>, IntoIter<(Position, FullPositionBreakdown)>, where Position comes two times and position id three time in the zip.
Also we can already make sure that zipped values of two tuple  (Position, Uint128, Uint128), (Position, FullPositionBreakdown)  are already sorted and have same position id on same index.

Reply:
> Also we can already make sure that zipped values of two tuple (Position, Uint128, Uint128), (Position, FullPositionBreakdown) are already sorted and have same position id on same index.
This is already checked in
```
            if pos1.position_id != pos2.position_id {
                // this would be a programming error
                panic!("Position ids in the zipped iterator don't match")
            }
```
We never hit this panic with the current code
Sorting again is not needed right now so would be just extra work

Reply:
>Could this be optmized in some way, as type of position_funds would become, Zip<Iter<(Position, Uint128, Uint128)>, IntoIter<(Position, FullPositionBreakdown)>, where Position comes two times and position id three time in the zip.

We could minimize this in a different struct, It would not lower the copy load afaik, but would be better legible. What do you think would work best here?